power: device: Add wakeup source API

ceolin opened the pull request:
Introduce a new API to allow devices capable of wake up the system
register themselves was wake up sources. This permits applications to
select the most appropriate way to wake up the system when it is
suspended.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Add wakeup source API

Introduce a new API to allow devices capable of wake up the system
register themselves was wake up sources. This permits applications to
select the most appropriate way to wake up the system when it is
suspended.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on include/pm/device.h:
```diff
@@ -121,6 +121,48 @@ struct pm_device {
+/**
+ * @brief Power management wakeup source struct
+ *
+ * Wakeup sources are used for controlling whether
+ * or not the device should use its system wakeup mechanism
+ * to wakeup the system. Devices capable of wakup the system can
+ * register their wakeup sources using @ref pm_wakeup_source_register
+ * to applications be able to use them
```
Can this be clarified?

erwango commented on include/pm/device.h:
```diff
@@ -121,6 +121,48 @@ struct pm_device {
+	struct k_spinlock lock;
+
+	/**
+	 * Device which the wakeup source belongs.
```
Device owner of the wake source

erwango commented on include/pm/device.h:
```diff
@@ -121,6 +121,48 @@ struct pm_device {
+	bool (*disable)(struct device *dev);
+
+	/**
+	 * Whether or the the wakeup source is enabled.
```
Wakeup source status flag

erwango commented on include/pm/device.h:
```diff
@@ -168,6 +210,73 @@ int pm_device_state_get(const struct device *dev, uint32_t *device_power_state);
+/**
+ * @brief Register a power management wakeup source
+ *
+ * Wakeup sources are used for controlling whether
+ * or not the device should use its system wakeup mechanism
+ * to wakeup the system. This function just register a wakeup
+ * source in the system, to enable the wakeup source use
+ * @ref pm_wakeup_source_enable
```
"Populates the provided wakeup source in system's wakeup source list" ?

erwango commented on include/pm/device.h:
```diff
@@ -168,6 +210,73 @@ int pm_device_state_get(const struct device *dev, uint32_t *device_power_state);
+/**
+ * @brief Disable a power management wakeup source
+ *
+ * Disable the given wakeup source, the wakeup source will be
+ * disabled and it will no longer be used to wake up the system.
```
```suggestion
 * Disable the given wakeup source: the wakeup source will no
 * no longer be used to wake up the system.
```

erwango reviewed (dismissed):
Edits for clarity.

One question on this API: should we consider affected power states for each wakeup source ?
For instance some wakeup sources could have effect in STOP mode but not in standby. Should it be visible from this API.

I guess this can have a significant impact on complexity, so it has to be added only if required, but I'm not clear if this should be visible from PM system.

gmarull commented:
Meeting notes (07/01/2021):

- Agreed to make the API _similar_ to what Linux offers (ref. https://www.kernel.org/doc/html/latest/driver-api/pm/devices.html#subsystem-level-methods)
- Add support for devicetree (ref. https://www.kernel.org/doc/Documentation/devicetree/bindings/power/wakeup-source.txt)

ceolin commented:
> Meeting notes (07/01/2021):
> 
> * Agreed to make the API _similar_ to what Linux offers (ref. https://www.kernel.org/doc/html/latest/driver-api/pm/devices.html#subsystem-level-methods)
> * Add support for devicetree (ref. https://www.kernel.org/doc/Documentation/devicetree/bindings/power/wakeup-source.txt)

Maybe it is not clear for me. The similarity with Linux will just be in how to declare a wake source in DTS and enable / disable it.  Zephyr's PM subsystem will look devices that are declared (and has enabled) wake source capability and account it. Devices drivers don't need to anything but declare this capability in DTS.

Commit:
pm: device: Do not suspend wake up sources

Check if a device is being used as a wake up source before suspend it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on subsys/pm/device.c:
```diff
@@ -151,3 +151,37 @@ int pm_device_state_get(const struct device *dev,
 }
+
+
+bool pm_device_wakeup_enable(struct device *dev, bool enable)
+{
+	atomic_val_t flags, new_flags;
+
+
+	flags =  atomic_get(&dev->pm->atomic_flags);
+
+	if ((flags & BIT(PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT)) == 0U) {
+		return false;
+	}
+
+	if (enable) {
+		new_flags = flags |
+			BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT);
+	} else {
+		new_flags = flags & ~BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT);
+	}
+
+	return atomic_cas(&dev->pm->atomic_flags, flags, new_flags);
+}
```
I'd personally simplify this function:

1. Make it `void` return type
2. Do nothing if the device is not wakeup capable (and document that this call will have no effect on non-wakeup capable devices)

Summarizing:
```c
void pm_device_wakeup_enable(struct device *dev, bool enable)
{
        if (!atomic_test_bit(&dev->pm->atomic_flags, PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT)) {
                return;
        }

        atomic_set_bit_to(&dev->pm->atomic_flags, PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT, enable);
}
```

ceolin replied:
I don't. If two threads call it simultaneously, one enabling and other disabling how would you know the result ? Using compare and set allows the caller to know if it succeed.

gmarull replied:
If two threads compete to do the opposite the last to do the operation will win, which I guess it's fine. Maybe I'm missing a usecase, though.

ceolin replied:
it is fine  but  who called it would no be able to know if it succeed.  I rather have it available.

gmarull replied:
If two threads are competing to do the opposite, it's an application design issue, not a failure. This API per se does not have any failure path when designed, e.g. with `atomic_set_bit_to`. If the application/device needs to check if the device is really enabled or not, it should call, e.g. `pm_device_wakeup_is_enabled` later.

ceolin replied:
the return of this function tells the caller whether or not it succeed also it returns false in case the device is not capable. We can eternally argue whether it is an application responsibility or not. I rather this way and it if it does not contain any flaw I would like to move with this implementation. Is it going to be a reason to block this pr ?

gmarull replied:
The problem is that `pm_device_wakeup_enable` can't fail as it is designed.

In case of Linux, for example, it can:
https://elixir.bootlin.com/linux/latest/source/drivers/base/power/wakeup.c#L336

If it needs to return a failure, then it's return type should be an `int` and return a negative errno. For example, if the device is not wakeup capable it could return `-EINVAL` as Linux does.

ceolin replied:
It can fail returning false. That means that operation did not succeed. Though, using an integer allows us to know what was the cause. Will change to that since you are pushing back for returning a boolean. But again, return false is ok for failing it tells that the operation didn't work.

ceolin replied:
btw, this kind of implies in changing `bool pm_device_wakeup_enable` to return `int` since it can fail because the device is not wake up capable. This api was just return false in case of failure and true in success. Is that really the path you wanna go ?

gmarull replied:
So how can a flag set fail? This is the part I don't really get.

ceolin replied:
fail if the device is not capable. You cannot enable a wakeup source if the device does no support it.

ceolin replied:
the way I'm using `atomi_cas` just ensure that no one else has changed this flags while this function is called, as for example if the device somehow loses the wake up capability. Of course, we are not covering it and just assuming that the never changes it is wakeup capability. So, use `atomic_set_bit_to` works but I tried to ensure that while I'm setting this flag no one else change these flags.

gmarull commented on include/device.h:
```diff
@@ -757,10 +757,19 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#define Z_DEVICE_STATE_DEFINE(node_id, dev_name)			\
+	static struct device_state Z_DEVICE_STATE_NAME(dev_name) = {	\
+		.pm = {							\
+			.atomic_flags = DT_PROP_OR(node_id, wakeup_source, 0), \
```
I think this can be directly initialized:
```c
.atomic_flags = ATOMIC_INIT((DT_PROP_OR(node_id, wakeup_source, 0) << PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT))
```

ceolin replied:
nice ! Will do it, was not worried about the value because this is properly set when the device is initialized, but have the proper value in the definition is more accurate.

gmarull commented on kernel/device.c:
```diff
@@ -32,6 +32,9 @@ static inline void device_pm_state_init(const struct device *dev)
 		.condvar = Z_CONDVAR_INITIALIZER(dev->pm->condvar),
+		.atomic_flags = dev->pm->atomic_flags ?
+			BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT) |
+		BIT(PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT) : 0,
```
see comment above

ceolin commented on dts/bindings/base/base.yaml:
```diff
@@ -95,3 +95,8 @@ properties:
       type: string-array
       description: Provided names of IO channel specifiers
+
+    wakeup-source:
+      required: false
+      type: boolean
+      description: It can be used as wake up source
```
we try to align properties names with Linux and there they use wakeup-source, unless this practice has changed I will keep `wakeup-source`.

gmarull replied:
Fair enough with the naming, but still I'd probably keep them in a separate file.

ceolin replied:
Indeed. We actually already have  a power.yaml to gather power related properties. Just moved it to there.

ceolin commented on include/device.h:
```diff
@@ -757,10 +757,19 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#define Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)	\
 	.pm_control = (pm_control_fn),				\
-	.pm = &Z_DEVICE_STATE_NAME(dev_name).pm,
+		.pm = &Z_DEVICE_STATE_NAME(dev_name).pm,
```
:/

ceolin commented on include/pm/device.h:
```diff
@@ -143,6 +155,42 @@ int pm_device_state_get(const struct device *dev,
+ *
+ * @param dev device object check.
+ *
+ * @return @c true if the device is wake up capable,
+ *         @c false otherwise.
```
really ? 
```
faceolin@faceolin-mobl5:~/p/zephyrproject/zephyr$ git grep "@retval" | wc -l
1424
faceolin@faceolin-mobl5:~/p/zephyrproject/zephyr$ git grep "@return " | wc -l
3677
```

gmarull replied:
I mean that `@retval` is used for each return value, eg

```
@retval true x
@retval false y
```

ceolin replied:
gotcha, but @return is more used than @retval. Do we have a guideline or a documentation about which one to use ? Because what you said about @retval being used in most places is not accurate.

gmarull replied:
`@retval` is used when specifying each return code. I can't find any `@return @c` in tree.

ceolin replied:
well, @c can be used anywhere. Most places just use `@return` followed by a value. I'm going to change it since I don't have a hard opinion about it.

gmarull reviewed (commented):
I won't block on the `pm_device_wakeup_is_enabled`, even though I'd personally not worry about the case it is trying to solve (Linux doesn't seem to care either, see https://elixir.bootlin.com/linux/latest/source/drivers/base/power/wakeup.c#L336).

Just left a few suggestions/questions. What should be added is a new doc page explaining the API, how it is expected to be used, how it behaves, etc.

ceolin commented on subsys/pm/device.c:
```diff
@@ -149,3 +149,36 @@ int pm_device_state_get(const struct device *dev,
+			       PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT);
+}
+
+bool pm_device_wakeup_is_capable(const struct device *dev)
```
Was wondering about it. Tried to keep it static, but don't see major problem to support this API.

ceolin replied:
lets keep it simpler without it since it is not clear whether or not it is necessary, Devices are not dynamically defined.

ceolin commented on kernel/device.c:
```diff
@@ -32,6 +32,10 @@ static inline void device_pm_state_init(const struct device *dev)
 		.usage = ATOMIC_INIT(0),
 		.lock = Z_MUTEX_INITIALIZER(dev->pm->lock),
 		.condvar = Z_CONDVAR_INITIALIZER(dev->pm->condvar),
+		.atomic_flags = dev->pm->atomic_flags ?
+		ATOMIC_INIT(BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT) |
+			    BIT(PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT)) :
+		ATOMIC_INIT(0),
```
> I won't block on the `pm_device_wakeup_is_enabled`, even though I'd personally not worry about the case it is trying to solve (Linux doesn't seem to care either, see https://elixir.bootlin.com/linux/latest/source/drivers/base/power/wakeup.c#L336).
> 
> Just left a few suggestions/questions. What should be added is a new doc page explaining the API, how it is expected to be used, how it behaves, etc.

That is not true, Linux returns `-EINVAL` if the device is not capable, we are returning `false` ...

gmarull replied:
> > I won't block on the `pm_device_wakeup_is_enabled`, even though I'd personally not worry about the case it is trying to solve (Linux doesn't seem to care either, see https://elixir.bootlin.com/linux/latest/source/drivers/base/power/wakeup.c#L336).
> > Just left a few suggestions/questions. What should be added is a new doc page explaining the API, how it is expected to be used, how it behaves, etc.
> 
> That is not true, Linux returns `-EINVAL` if the device is not capable, we are returning `false` ...

I mean the concurrency issues. Regarding return type, if we really want to signal errors we should likely use errnos.

ceolin commented:
@gmarull can I get an approval for this or is there anything else to review ?

gmarull commented:
I think that any new API should be accompanied by some docs (e.g. expected usage) and some sample/test that showcases its usage. I think this is particularly important for PM, since there are many PM issues in-tree in part because docs/examples are missing or incomplete.

Commit:
doc: pm: Add wakeup information

Add documentation for wakeup sources.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+-----------------
+
+Some devices are capable of wake the system up from a sleep state.
+Such devices have the bit :c:macro:`PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT` set
+in :c:member:`pm_device.atomic_flags`. When a device is "wakeup" capable, applications
```
These are implementation details that should be hidden from usage guidelines

gmarull replied:
```suggestion
Such devices have the bit :c:macro:`PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT` set
in :c:member:`pm_device.atomic_flags`. When a device has such capability, applications
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+Wakeup capability
+-----------------
+
+Some devices are capable of wake the system up from a sleep state.
```
```suggestion
Some devices are capable of waking the system up from a sleep state.
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+the device node in devicetree.  For example, this devicetree fragment sets
+``gpiol@40022000`` as "wakeup" capable:
+
+.. code-block:: devicetree
+
+		gpio0: gpio@40022000 {
+			compatible = "ti,cc13xx-cc26xx-gpio";
+			reg = <0x40022000 0x400>;
+			interrupts = <0 0>;
+			status = "disabled";
+			label = "GPIO_0";
+			gpio-controller;
+			wakeup-source;
+			#gpio-cells = <2>;
+		};
+
```
Considering that wakeup-sources are enabled by default, I don't think it is a good idea to place the flag in soc dts files, so I think:

- They need to be manually enabled, or
- The application sets the flag if needed, that is:
```
&gpio0 {
    wakeup-source;
};
```

erwango replied:
On my side, I'd say this is a good thing to be able to identify devices that are "wakeup source" capable in device tree.
Then, I'd suggest to disable them by default: treat them as other devices unless required by application.

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+
+This property can be set on device declaring the property ``wakeup-source`` in
+the device node in devicetree.  For example, this devicetree fragment sets
+``gpiol@40022000`` as "wakeup" capable:
```
```suggestion
the ``gpio0`` device as a wakeup source:
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+:c:func:`pm_device_wakeup_enable`.
+
+This property can be set on device declaring the property ``wakeup-source`` in
+the device node in devicetree.  For example, this devicetree fragment sets
```
```suggestion
the device node in devicetree. For example, this devicetree fragment sets
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -403,6 +403,45 @@ Check Busy Status of All Devices API
+   It is responsability of the application to do any additional configuration
+   required by the device to support it.
+
+
```
It would be nive to group wakeup API into a Doxygen group and list it here.

gmarull commented on kernel/device.c:
```diff
@@ -32,6 +32,7 @@ static inline void device_pm_state_init(const struct device *dev)
 		.usage = ATOMIC_INIT(0),
 		.lock = Z_MUTEX_INITIALIZER(dev->pm->lock),
 		.condvar = Z_CONDVAR_INITIALIZER(dev->pm->condvar),
+		.atomic_flags = ATOMIC_INIT(dev->pm->atomic_flags),
```
```suggestion
```

ceolin replied:
Nope, atomic_flags has to be initialized otherwise these devices will no be marked as wakeup capable.

gmarull replied:
So then, couldn't all these static initializers be moved to `Z_DEVICE_STATE_DEFINE`?

gmarull commented on include/device.h:
```diff
@@ -779,10 +779,21 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+		.pm = {						\
+			.atomic_flags = ATOMIC_INIT(DT_PROP_OR(	\
+				node_id, wakeup_source, 0) <<		\
+				PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT), \
```
Automatic enable is not happening now, ENABLE is missing here.

ceolin replied:
that is what is wanted right ? Not enable by default. Will keep it and change the documentation.

gmarull replied:
If we take the path of disabled by default yes, in which case this code is right. Docs need to be updated.

gmarull reviewed (changes requested):
Test/sample is still missing, added a few comments.

gmarull commented on dts/bindings/base/pm.yml:
```diff
@@ -0,0 +1,18 @@
+        wake up the system.
+
+        Wake up capable devices are enabled (interruptions will wake up
+        the system) by default but they can be disabled at runtime if necessary.
```
If `pm.yml` is not included by default, devices that use `wakeup-source` need to include it. Not sure what is better, but if not included by `base.yaml` docs should mention it.

ceolin replied:
This should be included in `base.yaml`, at least for now IMHO.

Commit:
drivers: gpio_emul: Add dummy pm support

Add a simple device PM support to be used for tests.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: pm: Add device wakeup test

Add a test to exercise PM device wakeup API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Change atomic flags type

Just using a simple atomic for flags instead of using an array.

While is neat using ATOMIC_DEFINE for future proof. The reality is
that it brings some problem for the wakeup source implementation
that needs to statically initialize it during the device definition.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on doc/_scripts/gen_devicetree_rest.py:
```diff
@@ -226,9 +226,11 @@ def load_base_binding():
     if not base_yaml.is_file():
         sys.exit(f'Expected to find base.yaml at {base_yaml}')
-    return edtlib.Binding(os.fspath(base_yaml), {}, require_compatible=False,
+    return edtlib.Binding(os.fspath(base_yaml), base_includes, require_compatible=False,
```
@mbolivar-nordic can you review these `gen_devicetree_rest.py` changes?

ceolin replied:
that is required because the base.yaml is including the new yaml in that directory.

mbolivar-nordic replied:
I think we should just add them to base.yaml, then.

mbolivar-nordic replied:
Never mind, I see you had a long discussion about this already that I didn't read at first. Sure; that seems fine as-is.

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -395,6 +395,44 @@ Check Busy Status of All Devices API
+
+This property can be set on device declaring the property ``wakeup-source`` in
+the device node in devicetree. For example, this devicetree fragment sets the
+``gpio`` device as a "wakeup" source:
```
```suggestion
``gpio0`` device as a "wakeup" source:
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -395,6 +395,44 @@ Check Busy Status of All Devices API
+
+   This property is ``only`` used by the system power management to identify
+   devices that should not be suspended.
+   It is responsability of the application to do any additional configuration
```
```suggestion
   It is responsability of the driver or the application to do any additional configuration
```

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -395,6 +395,44 @@ Check Busy Status of All Devices API
+
+.. note::
+
+   This property is ``only`` used by the system power management to identify
```
```suggestion
   This property is **only** used by the system power management to identify
```

gmarull commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -0,0 +1,117 @@
+		/* Enable wakeup source. Next time the system is called
+		 * to sleep, this device will still be active.
+		 */
+		(void)pm_device_wakeup_enable((struct device *)dev, true);
```
is cast to non-const necessary?

ceolin replied:
It is.

gmarull commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -0,0 +1,117 @@
+	ret = pm_device_wakeup_is_capable(dev);
+	zassert_true(ret, "Device marked as capable");
+
+	ret = pm_device_wakeup_enable((struct device *)dev, true);
```
is cast to non-const necessary?

ceolin replied:
ditto

gmarull commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -0,0 +1,117 @@
+	ret = pm_device_wakeup_is_enabled(dev);
+	zassert_true(ret, "Wakeup source not enabled");
+
+	ret = pm_device_wakeup_enable((struct device *)dev, false);
```
is cast to non-const necessary?

ceolin replied:
yep

mbolivar-nordic reviewed (approved):
for gen_devicetree_rest.py only

mbolivar-nordic commented on include/device.h:
```diff
@@ -732,10 +732,21 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#define Z_DEVICE_STATE_DEFINE(node_id, dev_name)			\
+	static struct device_state Z_DEVICE_STATE_NAME(dev_name) = {	\
+		.pm = {						\
+			.flags = ATOMIC_INIT(DT_PROP_OR(		\
+				node_id, wakeup_source, 0) <<		\
```
Just use DT_PROP.

ceolin replied:
nope, this does not work when something that expand to an invalid node. e.g: ``SYS_DEVICE_DEFINE("sys_clock"```

mbolivar-nordic replied:
> nope, this does not work when something that expand to an invalid node. e.g: ``SYS_DEVICE_DEFINE("sys_clock"```

Then I think the caller can do

```
IF_ENABLED(DT_NODE_EXISTS(node_id), (Z_DEVICE_STATE_DEFINE(node_id, dev_name))) \
```

or so

ceolin replied:
That implies in changes that are beyond this patch IMHO. The state has to be defined for cases where a device is not defined using DT. Of course, I can just do `IF_ENABLED(DT_NOT_EXISTS(node_id), ...) ` in that place, but it makes it much more verbose. Do you think is really to bad just keep this way ?

gmarull replied:
I agree with @mbolivar-nordic, and actually the same technique seems to be used for the device handles, see https://github.com/zephyrproject-rtos/zephyr/blob/0d3c75c2b09f9d11c04f52a92c5fb0fbef2ea32b/include/device.h#L696

Another change request I missed: we have `Z_DEVICE_DEFINE_PRE` for all stuff that needs to be defined before device, so I think the state definition also belongs to that group.

mbolivar-nordic replied:
>  Do you think is really to bad just keep this way ?

Yes, I am pushing everyone to use DT_PROP() instead of DT_NODE_HAS_PROP() when it comes to boolean properties for a reason. The reason is that using DT_PROP() means you benefit from the property system and type checking in the bindings language. If you use DT_NODE_HAS_PROP() instead, and you misspell the property, you can get an incorrect result and it would be very hard to debug. If you make a typo on the property name using DT_PROP(), then you get a build error instead. So using DT_NODE_HAS_PROP() for booleans is an antipattern I'm willing to block on.

ceolin replied:
the device handles is not doing the same thing, it is not accessing properties. Using `COND_CODE_1(DT_NODE_EXISTS(node_id), () ()` there is just more cumbersome than using `DT_PROP_OR` IMHO, since they are doing exactly the same thing. Now ... if you are saying to *NOT* define the state struct in case the device is not defined using DT that is not right because this struct is required for devices that are defined manually using `DEVICE_DEFINE`.

So in the end, it seems just a matter of changing  `DT_PROP_OR` with `COND_CODE_1(DT_NODE_EXISTS(node_id) ...`. I particularly prefer the first, it looks cleaner in my opinion, but I am ok changing it.

Regards `Z_DEVICE_DEFINE_PRE`, that makes sense.

mbolivar-nordic reviewed (changes requested):
Found at least one code nit so I'm removing my +1

mbolivar-nordic commented on include/pm/device.h:
```diff
@@ -66,6 +66,13 @@ enum pm_device_state {
+	/**
+	 * Indicates whether or not the device is capable of waking the system
+	 * up.
+	 **/
```
Non-blocking comment style nit here.

ceolin replied:
why ? this is consistent with comments in this file. The comment does not fit in one line.

mbolivar-nordic replied:
> why ? this is consistent with comments in this file. The comment does not fit in one line.


```
	 */
```

not

```
	 **/
```

ceolin replied:
have changed it ;)

mbolivar-nordic commented on include/pm/device.h:
```diff
@@ -231,6 +238,42 @@ __deprecated static inline int device_busy_check(const struct device *dev)
+/**
+ * @brief Check if a device is wake up capable
+ *
+ * @param dev device object check.
```
s/check/to check/

mbolivar-nordic commented on doc/reference/power_management/index.rst:
```diff
@@ -395,6 +395,44 @@ Check Busy Status of All Devices API
+			#gpio-cells = <2>;
+		};
+
+By default, "wakeup" capable devices are not enabled
+during the device initialization. Applications can enable this
+functionality later calling :c:func:`pm_device_wakeup_enable`.
```
This is confusing to me as an outside reader. It kind of sounds like you're saying the device will not be initialized. I don't know what "enabling" a device "during the device initialization" means except for initializing it. What am I missing here? Is there some way to make it clearer?

Not a blocking comment, but it would be nice to get some clarification here.

ceolin replied:
that makes sense. I've read it again and it is indeed confuse.

Will change to;
```
... capable devices do not have this functionality enabled during the device initialization....
```

mbolivar-nordic commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -0,0 +1,117 @@
+{
+	bool ret = false;
+
+	dev = device_get_binding(DEV_NAME);
```
Just use DEVICE_DT_GET. There's no reason anymore to be using device_get_binding for devices that are always set up in devicetree.

mbolivar-nordic reviewed (changes requested):
A couple of other things I noticed while looking this over.

Public service announcement regarding boolean properties from the `DT_PROP` documentation:

```
 * For properties whose bindings have the following types, this macro
 * expands to:
 * [...]
 * - boolean: 0 if the property is false, or 1 if it is true
```

There's no need to use DT_NODE_HAS_PROP or DT_PROP_OR with boolean properties. I see this happen a lot and it's not necessary. The type system knows what to do with these; using DT_PROP is safe.

ceolin commented:
> A couple of other things I noticed while looking this over.
> 
> Public service announcement regarding boolean properties from the `DT_PROP` documentation:
> 
> ```
>  * For properties whose bindings have the following types, this macro
>  * expands to:
>  * [...]
>  * - boolean: 0 if the property is false, or 1 if it is true
> ```
> 
> There's no need to use DT_NODE_HAS_PROP or DT_PROP_OR with boolean properties. I see this happen a lot and it's not necessary. The type system knows what to do with these; using DT_PROP is safe.

The problem is when you have something that expand to an  invalid node. e.g:  ``SYS_DEVICE_DEFINE("sys_clock"```
`DT_PROP_OR` will expand to the default value while `DT_PROP` will break.

ceolin commented:
@mbolivar @gmarull  Can you folks review it again ?
